sql/core/src/test/scala/org/apache/spark/sql/streaming/continuous/ContinuousSuite.scala

```diff
@@ -38,18 +38,42 @@ class ContinuousSuiteBase extends StreamTest {
       sparkConf.set("spark.sql.testkey", "true")))
 
   protected def waitForRateSourceTriggers(query: StreamExecution, numTriggers: Int): Unit = {
-    query match {
+    findRateStreamContinuousStream(query).foreach { reader =>
+      val deltaMs = numTriggers * 1000 + 300
+      val firstCommittedTime = reader.firstCommittedTime.longValue()
+      while (System.currentTimeMillis < firstCommittedTime + deltaMs) {
+        Thread.sleep(firstCommittedTime + deltaMs - System.currentTimeMillis)
+      }
+    }
+  }
+
+  protected def waitForRateSourceCommittedValue(
```

Comment:
This is safest approach to expect some rows to produce outputs. We still need to have max time to wait, since it may block infinitely in case of bugs.